vllm/v1/sample/sampler.py

```diff
@@ -132,6 +183,12 @@ def sample(
             logits = processor.apply(logits)
 
         # Apply top_k and/or top_p.
+        if return_final_logprobs:
+            # Apply top_k and/or top_p here to avoid unnecessary computation
+            # in `TopKTopPSampler` efficient implementations.
+            final_logprobs = self.compute_logprobs(
+                apply_top_k_top_p(logits.clone(), sampling_metadata.top_k,
+                                  sampling_metadata.top_p))
```

Comment:
Shouldn't this be handled inside `topk_topp_sampler`? IIUC, we are computing this twice if `return_final_logprobs = True`.

Reply:
Can you re-arrange the sampler somehow to avoid this? cc @njhill

Reply:
One way is to compute `logsoftmax` first and save it for return, then do `exp`, because logprobs accuracy is more important than probs. Shall we take this approach? Note even with this, we cannot avoid the full re-computation (top_k, top_p, logsoftmax) for the flashinfer path.

Reply:
I think we should use `apply_top_k_top_p` (instead of flashinfer) if the logprobs are required. If we already applied top-p and top-k masking to the logits, we don't need to use flashinfer's sampling kernel.

Reply:
My suggestion is to restructure the code so that
* when processed logprobs are required, compute top-p/top-k mask only once and use it for both (non-flashinfer) sampling and logprobs
* Otherwise, just use flashinfer (if it's available) for sampling

Reply:
Making the change, will update